[flang][OpenMP] Upstream first part of `do concurrent` mapping (#126026)
llvm/llvm-project: flang/docs/DoConcurrentConversionToOpenMP.md

as well.

### More advanced detection of loop nests

As pointed out earlier, any intervening code between the headers of 2 nested
`do concurrent` loops prevents us currently from detecting this as a loop nest.
In some cases this is overly conservative. Therefore, a more flexible detection
logic of loop nests needs to be implemented.

### Data-dependence analysis

Right now, we map loop nests without analysing whether such mapping is safe to
do or not. We probably need to at least warn the use of unsafe loop nests due
to loop-carried dependencies.

### Non-rectangular loop nests

So far, we did not need to use the pass for non-rectangular loop nests. For
example:
```fortran
do concurrent(i=1:n)
  do concurrent(j=i:n)
    ...
  end do
end do
```
We defer this to the (hopefully) near future when we get the conversion in a
good share for the samples/projects at hand.

### Generalizing the pass to other parallel programming models

Once we have a stable and capable `do concurrent` to OpenMP mapping, we can take
this in a more generalized direction and allow the pass to target other models;
e.g. OpenACC. This goal should be kept in mind from the get-go even while only
targeting OpenMP.

> 

Reviewer: If reductions are not handled can you mention that here?

   SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

-->

# `DO CONCURRENT` mapping to OpenMP

```{contents}
---
local:
---
```

This document seeks to describe the effort to parallelize `do concurrent` loops
by mapping them to OpenMP worksharing constructs. The goals of this document
are:
* Describing how to instruct `flang` to map `DO CONCURRENT` loops to OpenMP
  constructs.
* Tracking the current status of such mapping.
* Describing the limitations of the current implementation.
* Describing next steps.
* Tracking the current upstreaming status (from the AMD ROCm fork).

## Usage

In order to enable `do concurrent` to OpenMP mapping, `flang` adds a new
compiler flag: `-fdo-concurrent-to-openmp`. This flag has 3 possible values:
1. `host`: this maps `do concurrent` loops to run in parallel on the host CPU.
   This maps such loops to the equivalent of `omp parallel do`.
2. `device`: this maps `do concurrent` loops to run in parallel on a target device.
   This maps such loops to the equivalent of
   `omp target teams distribute parallel do`.
3. `none`: this disables `do concurrent` mapping altogether. In that case, such
   loops are emitted as sequential loops.

> The above compiler switch is currently available only when OpenMP is also

Reviewer 2: NIT: Consider replacing above with the actual compiler switch. It feels slightly more readable to me.

```suggestion
The `-fdo-concurrent-to-openmp` compiler switch is currently available only when OpenMP is also
```

OpenMP is also enabled. So you need to provide the following options to flang in
order to enable it:
```
flang ... -fopenmp -fdo-concurrent-to-openmp=[host|device|none] ...
```
For mapping to device, the target device architecture must be specified as well.

> See `-fopenmp-targets` and `-foffload-arch` for more info.

Reviewer 3: ```suggestion
See `-fopenmp-targets` and `--offload-arch` for more info.
```

At the moment, the FIR dialect does not have a way to model locality specifiers
on the IR level. Instead, something similar to early/eager privatization in OpenMP
is done for the locality specifiers in `fir.do_loop` ops. Having locality specifier
modelled in a way similar to delayed privatization (i.e. the `omp.private` op) and
reductions (i.e. the `omp.declare_reduction` op) can make mapping `do concurrent`
to OpenMP (and other parallel programming models) much easier.

Therefore, one way to approach this problem is to extract the TableGen records
for relevant OpenMP clauses in a shared dialect for "data environment management"
and use these shared records for OpenMP, `do concurrent`, and possibly OpenACC
as well.

#### Supporting reductions

Similar to locality specifiers, mapping reductions from `do concurrent` to OpenMP
is also still an open TODO. We can potentially extend the MLIR infrastructure
proposed in the previous section to share reduction records among the different 
relevant dialects as well.

### More advanced detection of loop nests

As pointed out earlier, any intervening code between the headers of 2 nested

> `do concurrent` loops prevents us currently from detecting this as a loop nest.

Reviewer 2: NIT

```suggestion
`do concurrent` loops prevents us from detecting this as a loop nest.
```

In some cases this is overly conservative. Therefore, a more flexible detection
logic of loop nests needs to be implemented.

### Data-dependence analysis

Right now, we map loop nests without analysing whether such mapping is safe to

> do or not. We probably need to at least warn the use of unsafe loop nests due

Reviewer 2: NIT: Did you mean "user" here?

```suggestion
do or not. We probably need to at least warn the user of unsafe loop nests due
```

<!--===- docs/DoConcurrentMappingToOpenMP.md

   Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
   See https://llvm.org/LICENSE.txt for license information.
   SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

-->

> # `DO CONCURENT` mapping to OpenMP

Reviewer: ```suggestion
# `DO CONCURRENT` mapping to OpenMP
```

```{contents}
---
local:
---
```

This document seeks to describe the effort to parallelize `do concurrent` loops
by mapping them to OpenMP worksharing constructs. The goals of this document
are:

> * Describing how to instruct `flang` to map `DO CONCURENT` loops to OpenMP

Reviewer: ```suggestion
* Describing how to instruct `flang` to map `DO CONCURRENT` loops to OpenMP
```

  constructs.
* Tracking the current status of such mapping.
* Describing the limitations of the current implmenentation.
* Describing next steps.
* Tracking the current upstreaming status (from the AMD ROCm fork).

## Usage

In order to enable `do concurrent` to OpenMP mapping, `flang` adds a new
compiler flag: `-fdo-concurrent-to-openmp`. This flag has 3 possible values:

> 1. `host`: this maps `do concurent` loops to run in parallel on the host CPU.

Reviewer: ```suggestion
1. `host`: this maps `do concurrent` loops to run in parallel on the host CPU.
```

   This maps such loops to the equivalent of `omp parallel do`.

> 2. `device`: this maps `do concurent` loops to run in parallel on a target device.

Reviewer: ```suggestion
2. `device`: this maps `do concurrent` loops to run in parallel on a target device.
```

   This maps such loops to the equivalent of
   `omp target teams distribute parallel do`.
3. `none`: this disables `do concurrent` mapping altogether. In that case, such
   loops are emitted as sequential loops.

The above compiler switch is currently available only when OpenMP is also
enabled. So you need to provide the following options to flang in order to
enable it:
```

> flang ... -fopenmp -fdo-concurrent-to-openmp=[host|device|none] ...

Reviewer: Is `-fopenmp-targets` required for the device?

Author: Yes, I expanded this point to clarify this.

OpenMP region similar to what the single-range example in previous section
shows.

#### Multi-range and perfectly-nested loops

Currently, on the `FIR` dialect level, the following loop:
```fortran
do concurrent(i=1:n, j=1:m)
  a(i,j) = i * j
end do
```
is modelled as a nest of `fir.do_loop` ops such that the outer loop's region
contains:
  1. The operations needed to assign/update the outer loop's induction variable.
  1. The inner loop itself.

So the MLIR structure looks similar to the following:
```
fir.do_loop %arg0 = %11 to %12 step %c1 unordered {
  ...
  fir.do_loop %arg1 = %14 to %15 step %c1_1 unordered {
    ...
  }
}
```
This applies to multi-range loops in general; they are represented in the IR as
a nest of `fir.do_loop` ops with the above nesting structure.

Therefore, the pass detects such "perfectly" nested loop ops to identify multi-range
loops and map them as "collapsed" loops in OpenMP.

#### Further info regarding loop nest detection

Loop-nest detection is currently limited to the scenario described in the previous
section. However, this is quite limited and can be extended in the future to cover

> more cases. For example, for the following loop nest, even thought, both loops are
> perfectly nested; at the moment, only the outer loop is parallized:

Reviewer 3: ```suggestion
more cases. For example, for the following loop nest, even though both loops are
perfectly nested, at the moment only the outer loop is parallelized:
```

```fortran
do concurrent(i=1:n)
  do concurrent(j=1:m)
    a(i,j) = i * j
  end do
end do

> ```

Reviewer: Is that for multi-range loops as well?

Similarly, for the following loop nest, even though the intervening statement `x = 41`
does not have any memory effects that would affect parallelization, this nest is
not parallelized as well (only the outer loop is).

```fortran
do concurrent(i=1:n)
  x = 41
  do concurrent(j=1:m)
    a(i,j) = i * j
  end do
end do

> ```

Reviewer: Just remembered a case here. Do we privatize the `j` loop index when converting the do concurrent to `parallel + omp do`?
```
do concurrent(i=1:n)
  do j=1,m
    a(i,j) = i * j
  end do
end do
``

Author: No, it will be shared. This is described in the next paragraph. At least this how the downstream implementation currently works. We can of course descide something when I upstream the next part.

Reviewer 4: I would read p 201 of ISO/IEC 1539:1:2023 that `j` should be privatized unless it has the SAVE attribute.  I will send a note to J3/WG5 and make sure that I'm following the right logical path when reading the Fortran spec.

Author: We can re-open the discussion when the relevant part is upstreamed. This part of the doc was removed until later in any case.

The above also has the consequence that the `j` variable will **not** be
privatized in the OpenMP parallel/target region. In other words, it will be
treated as if it was a `shared` variable. For more details about privatization,
see the "Data environment" section below.

See `flang/test/Transforms/DoConcurrent/loop_nest_test.f90` for more examples
of what is and is not detected as a perfect loop nest.

### Data environment

By default, variables that are used inside a `do concurrent` loop nest are
either treated as `shared` in case of mapping to `host`, or mapped into the
`target` region using a `map` clause in case of mapping to `device`. The only
exceptions to this are:
  1. the loop's iteration variable(s) (IV) of **perfect** loop nests. In that

>      case, for each IV, we allocate a local copy as shown the by the mapping

Reviewer: ```suggestion
     case, for each IV, we allocate a local copy as shown by the mapping
```

     examples above.
  1. any values that are from allocations outside the loop nest and used
     exclusively inside of it. In such cases, a local privatized

>      value is created in the OpenMP region to prevent multiple teams of threads

Reviewer 3: Nit: I think this is easier to understand, but feel free to ignore.
```suggestion
     copy is created in the OpenMP region to prevent multiple teams of threads
```

>      from accessing and destroying the same memory block which causes runtime

Reviewer 3: ```suggestion
     from accessing and destroying the same memory block, which causes runtime
```

     issues. For an example of such cases, see
     `flang/test/Transforms/DoConcurrent/locally_destroyed_temp.f90`.

> Implicit mapping detection (for mapping to the GPU) is still quite limited and

Reviewer 3: ```suggestion
Implicit mapping detection (for mapping to the target device) is still quite limited and
```

work to make it smarter is underway for both OpenMP in general and `do concurrent`
mapping.

#### Non-perfectly-nested loops' IVs

For non-perfectly-nested loops, the IVs are still treated as `shared` or
`map` entries as pointed out above. This **might not** be consistent with what
the Fortran specficiation tells us. In particular, taking the following
snippets from the spec (version 2023) into account:

> § 3.35
> ------
> construct entity
> entity whose identifier has the scope of a construct

> § 19.4
> ------
>  A variable that appears as an index-name in a FORALL or DO CONCURRENT

> >  construct, or ... is a construct entity. A variable that has LOCAL or

Reviewer 3: ```suggestion
>  construct [...] is a construct entity. A variable that has LOCAL or
```

>  LOCAL_INIT locality in a DO CONCURRENT construct is a construct entity.

> > ...

Reviewer 3: ```suggestion
> [...]
```
